### PR TITLE
Add parameters to limit number of settled nodes during preparation, much faster preparation for certain graphs

### LINKED ISSUE #34: Why does the preparation time vary so much for different maps?
The preparation time varies drastically depending on the graph we use. Take for example these graphs (included in this repository), that are similar in size:

|name|nodes|edges|edges/nodes|preparation time (ms)|fast_graph edges|fast_graph edges/node|prep per node (μs)|query time (μs)|
|-|-|-|-|-|-|-|-|-|
|Bremen dist|40_461|85_090|2.10|429|66_520|1.64|10|18|
|Bremen time|40_461|85_090|2.10|6_868|62_518|1.54|169|13|
|Bremen uniform*|40_461|85_090|2.10|307|65_669|1.62|8|15|
|South Seattle|29_763|49_604|1.67|16_750|64_029|2.15|560|32|

The measurements shown in the table are just the result of running the tests in `lib.rs` on my laptop. 

*Update: For the Bremen uniform graph I used the Bremen time graph, but set all edge weights to a constant value of 10.

I'd really like to find out where this difference comes from and it would be especially useful to speed up the preparation of the abstreet graphs (South Seattle here) somehow. Also related: #33.

### PR DESCRIPTION
**Pull request by easbar**
Fixes #34.

Long story short: When there are large-weight edges the preparation time can become much slower because witness searches explore too many nodes. Preventing this and still not adding too many shortcuts is non-trivial, so this PR introduces some parameters that can be used to tune performance depending for different types of graphs. It also provides some default values that should work reasonably well in most cases.

todo: 

* [x] update benchmark values in README.md
* [x] add benchmark values to mini-performance tests in `lib.rs` and maybe further tune parameters

### CODE REVIEW

File: src/fast_graph_builder.rs [hunk 1]
```diff
@@ -242,18 +271,63 @@ impl FastGraphBuilder {
 pub struct Params {
     pub hierarchy_depth_factor: f32,
     pub edge_quotient_factor: f32,
+    /// The maximum number of settled nodes per witness search performed when priorities are
+    /// calculated for all nodes initially. Since this does not take much time normally you should
+    /// probably keep the default.
+    pub max_settled_nodes_initial_relevance: usize,
+    /// The maximum number of settled nodes per witness search performed when updating priorities
```
Comment (dabreegster) [inline src/fast_graph_builder.rs:286]: These explanations are great, thank you! I kind of want to ask for something similar for `hierarchy_depth_factor` and `edge_quotient_factor` -- would users ever want to deviate from defaults?

**Review by dabreegster — APPROVED** [review]
Awesome, thanks!

File: src/fast_graph_builder.rs (hunk 1 above)
Comment (easbar) [reply src/fast_graph_builder.rs:286]: Good point, I added some documentation for these as well. I also ran some measurements for varying values of `hierarchy_depth_factor`, see here: https://gist.github.com/easbar/a7980e834206b6065e848ba29e228fdd#hierarchy-depth-factor

For the abstreet maps you sent reducing `hierarchy_depth_factor` from `0.1` to `0.01` might actually reduce preparation times further (but also might yield somewhat slower queries, ~10% according to these experiments, but that could also just be noise).

**Comment by easbar** [comment]
> That sounds unexpected for bike/car/walk. Most nodes are directed road segments, and from a quick glance at a reachability debugger, there are only a few roads not reachable:

Hm, strange. I might also check if the graphs are strongly connected. I agree that looking at your reachability debugger 25% of failed paths are unexpected.

> The new default settings are indeed a huge performance boost.

I'm glad to hear this. Note that for your maps I think `Params::new(0.01, 100, 10, 100)` and `ParamsWithOrder::new(100)` work better than the defaults. I'll send a PR once this is merged.

> For now, there's no recorded set of realistic user edits, but I'll keep an eye on the re-preparation times as I test things.

Yes, a set of realistic edits would be helpful to analyze this further for sure.

**Pull request merged**